### PR TITLE
Feature/670 Implement ZK Circuit using Snarkjs

### PR DESCRIPTION
**Pull request by biscuitdey**
# Description

This PR sets up the basic interfaces for the zk circuit agent, service and snarkjs library. As the creation and verification of zk proof is done internally (within a workstep), I have not added any API url in the controller.

## Related Issue

#670 

## How Has This Been Tested

This is a simple interface. Test is not applicable.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts [hunk 1]
```diff
@@ -5,10 +5,10 @@ import { SnarkjsCircuitService } from './snarkjs/snarkjs';
 export class CircuitService {
   constructor(private readonly circuit: SnarkjsCircuitService) {}
   async createProof(inputs: object) {
-    return { proof: 'proof', publicInput: 'publicInput' };
+    return await this.circuit.createProof(inputs);
   }
 
   async verifyProof(proof, publicInputs): Promise<boolean> {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts]: add param types

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts [hunk 2]
```diff
@@ -1,12 +1,32 @@
+import snarkjs from 'snarkjs';
 export class SnarkjsCircuitService {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts:2]: missing @Injectable

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts [hunk 3]
```diff
@@ -1,12 +1,32 @@
...
   public async createProof(inputs: object) {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts]: missing return type

File: examples/bri-3/types/snarkjs.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+declare module 'snarkjs';
```
Comment (ognjenkurtic) [inline examples/bri-3/types/snarkjs.d.ts]: why is this needed?

File: examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export class Witness {
+  proof: any;
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts:2]: does it have to be any? can it be string?

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@nestjs/common';
+import { CircuitService } from '../services/circuit/circuit.service';
+import { Witness } from '../models/witness';
+
+@Injectable()
+export class CircuitAgent {
+  constructor(private readonly circuitAgent: CircuitService) {}
+
+  public async createProof(input: any): Promise<Witness> {
+    const witness = await this.circuitAgent.createProof(input);
```
Comment (skosito) [inline examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts:10]: small thing but you can just `return this.circuitAgent ...` same below

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { Injectable } from '@nestjs/common';
+import { SnarkjsCircuitService } from './snarkjs/snarkjs';
```
[the remark this answers is not in the document]
Comment (skosito) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts:2]: agree, seems some of these are redundant, and maybe even one service/agent would be enough for now

File: examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts (hunk 5 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts:2]: In snarkjs, Proof is a json object with this format
 
 proof:  {
      pi_a: [
        '287015515026886850367000370304282418506884049051461050087416899760168881497',
        '18148864089385570498905488155229676521878972897521971323742422980058890911751',
        '1'
      ],
      pi_b: [
        [
          '4105642086103088183689535058970939306131735480592400919288550096864124823539',
          '9073470526838081637126023810680740679551786524097096458194988998006765175875'
        ],
        [
          '478070075091370606229554665611856500549347842347824569455531309060744385612',
          '17618482729486526802793459583506757327991535735572167103342901581209927978239'
        ],
        [ '1', '0' ]
      ],
      pi_c: [
        '9844961605053025613073899766400828870655794410013247320323691464876851190592',
        '14679854688986179956949270074292341730632138546772235800957114543397279031062',
        '1'
      ],
      protocol: 'groth16',
      curve: 'bn128'
    }

    Gnark also returns the pi_a, pi_b, pi_c properties but with different naming conventions.

File: examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
+  publicInput?: any;
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/models/witness.ts:4]: In snarkjs, it is an array of strings. I am not sure how other libraries might treat it. Hence, I kept it as any.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+@Injectable()
+export class CircuitAgent {
+  constructor(private readonly circuitAgent: CircuitService) {}
+
+  public async createProof(): Promise<any> {
+    const { proof, publicInput } = await this.circuitAgent.createProof();
+    return { proof, publicInput };
+  }
+
+  public async verifyProof(proof, publicInput): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { Injectable } from '@nestjs/common';
+import { CircuitService } from '../services/circuit/circuit.service';
+
+@Injectable()
+export class CircuitAgent {
+  constructor(private readonly circuitAgent: CircuitService) {}
+
+  public async createProof(): Promise<any> {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts (hunk 2 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts:2]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts:2]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts (hunk 2 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts:2]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { Injectable } from '@nestjs/common';
+import { CircuitService } from '../services/circuit/circuit.service';
+
+@Injectable()
+export class CircuitAgent {
+  constructor(private readonly circuitAgent: CircuitService) {}
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts:7]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts [hunk 12]
```diff
@@ -5,10 +5,10 @@ import { SnarkjsCircuitService } from './snarkjs/snarkjs';
 export class CircuitService {
   constructor(private readonly circuit: SnarkjsCircuitService) {}
   async createProof(inputs: object) {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts (hunk 1 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/circuit.service.ts]: Fixed.

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts (hunk 3 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/snarkjs.ts]: Fixed.

File: examples/bri-3/types/snarkjs.d.ts (hunk 4 above)
Comment (biscuitdey) [reply examples/bri-3/types/snarkjs.d.ts]: Yes, snarkjs is missing types. :(

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@nestjs/common';
+import { CircuitService } from '../services/circuit/circuit.service';
+import { Witness } from '../models/witness';
+
+@Injectable()
+export class CircuitAgent {
+  constructor(private readonly circuitAgent: CircuitService) {}
+
+  public async createProof(input: any): Promise<Witness> {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts:9]: I separated both into different functions.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts (hunk 6 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/circuit.agent.ts:10]: Fixed.

**Review by Kasshern — APPROVED** [review]
Very nice and clean start.

**Review by ognjenkurtic — DISMISSED** [review]

**Comment by ognjenkurtic** [comment]
Please fix conflicts and we can merge

File: examples/bri-3/src/bri/zeroKnowledgeProof/models/proof.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+export class Proof {
+  a: string[];
```
Comment (Manik-Jain) [inline examples/bri-3/src/bri/zeroKnowledgeProof/models/proof.ts:2]: not sure what does a,b,c signify. Either change these variable names, or add a comment about what they signify.

Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/models/proof.ts:9]: Fixed.

**Review by Kasshern — APPROVED** [review]

**Review by ognjenkurtic — APPROVED** [review]
